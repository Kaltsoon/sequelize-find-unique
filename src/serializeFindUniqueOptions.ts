import { ModelStatic } from 'sequelize/types';
import jsonStringify from 'json-stable-stringify';

import { isSequelizeModelClass } from './utils';
import { FindUniqueOptions } from './types';

interface Serializer {
  test: (key: string, value: unknown) => boolean;
  serialize: (key: string, value: unknown) => string;
}

const serializers: Serializer[] = [
  {
    test: (key: string, value: unknown) => isSequelizeModelClass(value),
    serialize: (key: string, value: unknown) =>
      `__model:${(value as ModelStatic<any>).getTableName()}`,
  },
];

const serializeReplacer = (key: string, value: unknown): unknown => {
  const serializer = serializers.find(({ test }) => test(key, value));

  if (serializer) {
    return serializer.serialize(key, value);
  }

  if (typeof value === 'function') {
    throw new Error('Found non-serializable function from findUnique options');
  }

  return value;
};

const serializeFindUniqueOptions = (
  options: FindUniqueOptions<any>,
): string => {
  const { where, include, attributes } = options;

  const whereKeys = where ? Object.keys(where) : [];

  return jsonStringify(
    { where: whereKeys, include, attributes },
    { replacer: serializeReplacer },
  );
};

export default serializeFindUniqueOptions;
