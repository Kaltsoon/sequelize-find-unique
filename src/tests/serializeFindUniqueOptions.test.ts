import { Op } from 'sequelize';

import serializeFindUniqueOptions from '../serializeFindUniqueOptions';
import { sequelize, Comment } from './models';

describe('serializeFindUniqueOptions', () => {
  test('serializes primitive values correctly', () => {
    expect(
      serializeFindUniqueOptions({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: ['firstName', 'lastName'],
      }),
    ).toMatchSnapshot();
  });

  test('serializes model references correctly', () => {
    expect(
      serializeFindUniqueOptions({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: ['firstName', 'lastName'],
        include: [{ model: Comment, as: 'comments' }],
      }),
    ).toMatchSnapshot();
  });

  test('serializes fn correctly', () => {
    expect(
      serializeFindUniqueOptions({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: [
          'firstName',
          'lastName',
          [sequelize.fn('COUNT', sequelize.col('column')), 'column_alias'],
        ],
      }),
    ).toMatchSnapshot();
  });

  test('serializes Op correctly', () => {
    expect(
      serializeFindUniqueOptions({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: ['firstName', 'lastName'],
        include: [
          {
            model: Comment,
            as: 'comments',
            where: {
              [Op.or]: [{ id: { [Op.gt]: 5 } }, { id: { [Op.ne]: 100 } }],
            },
          },
        ],
      }),
    ).toMatchSnapshot();
  });

  test('throws error for non-serializable values', () => {
    expect(() =>
      serializeFindUniqueOptions({
        where: {
          firstName: 'Kalle',
          lastName: 'Ilves',
        },
        attributes: [
          'firstName',
          'lastName',
          () => {
            return undefined;
          },
        ] as any,
      }),
    ).toThrowErrorMatchingSnapshot();
  });
});
