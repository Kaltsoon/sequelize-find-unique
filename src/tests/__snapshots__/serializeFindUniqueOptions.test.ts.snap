// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeFindUniqueOptions serializes Op correctly 1`] = `"{\\"attributes\\":[\\"firstName\\",\\"lastName\\"],\\"include\\":[{\\"as\\":\\"comments\\",\\"model\\":\\"__model:comments\\",\\"where\\":{\\"__op:Symbol(or)\\":[{\\"id\\":{\\"__op:Symbol(gt)\\":5}},{\\"id\\":{\\"__op:Symbol(ne)\\":100}}]}}],\\"where\\":[\\"firstName\\",\\"lastName\\"]}"`;

exports[`serializeFindUniqueOptions serializes fn correctly 1`] = `"{\\"attributes\\":[\\"firstName\\",\\"lastName\\",[{\\"args\\":[{\\"col\\":\\"column\\"}],\\"fn\\":\\"COUNT\\"},\\"column_alias\\"]],\\"where\\":[\\"firstName\\",\\"lastName\\"]}"`;

exports[`serializeFindUniqueOptions serializes model references correctly 1`] = `"{\\"attributes\\":[\\"firstName\\",\\"lastName\\"],\\"include\\":[{\\"as\\":\\"comments\\",\\"model\\":\\"__model:comments\\"}],\\"where\\":[\\"firstName\\",\\"lastName\\"]}"`;

exports[`serializeFindUniqueOptions serializes primitive values correctly 1`] = `"{\\"attributes\\":[\\"firstName\\",\\"lastName\\"],\\"where\\":[\\"firstName\\",\\"lastName\\"]}"`;

exports[`serializeFindUniqueOptions throws error for non-serializable values 1`] = `"Found non-serializable function from findUnique options"`;
